File: PitBoard(Android)/app/src/main/java/com/pitboard/app/util/ColorContrast.kt
package com.pitboard.app.util

import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.toArgb
import kotlin.math.abs

/**
 * Utilidades de contraste para pintar texto legible sobre colores de categoría arbitrarios
 * (el usuario los elige libremente como texto en CategoryManagerScreen, sin selector visual
 * ni vista previa de contraste). Esta lógica vivía solo dentro de RaceWidget.kt — el widget ya
 * protegía sus etiquetas, pero la app de teléfono pintaba el texto en blanco fijo encima de
 * cualquier color, así que un color claro podía dejar el texto invisible. Ahora la comparten
 * los dos sitios.
 */
object ColorContrast {
    fun perceivedLuminance(color: Color): Float =
        0.299f * color.red + 0.587f * color.green + 0.114f * color.blue

    /** Empuja [color] hacia claro/oscuro si su luminancia percibida está demasiado cerca de
     *  la de [background], para que no se confundan visualmente entre sí. */
    fun ensureContrast(color: Color, background: Color): Color {
        if (abs(perceivedLuminance(color) - perceivedLuminance(background)) >= 0.25f) return color
        val hsv = FloatArray(3)
        android.graphics.Color.colorToHSV(color.toArgb(), hsv)
        hsv[2] = if (perceivedLuminance(background) < 0.5f) {
            (hsv[2] + 0.4f).coerceAtMost(1f)
        } else {
            (hsv[2] - 0.4f).coerceAtLeast(0f)
        }
        return Color(android.graphics.Color.HSVToColor(hsv))
    }

    /** Blanco o negro, el que mejor se lea encima de [background]. */
    fun readableTextColor(background: Color): Color =
        if (perceivedLuminance(background) > 0.55f) Color.Black else Color.White

    fun safeParseColor(hex: String, fallback: Color = Color(0xFF5F6570)): Color = try {
        Color(android.graphics.Color.parseColor(hex))
    } catch (_: IllegalArgumentException) {
        fallback
    }
}

File: PitBoard(Iphone)/PitBoardKit/Sources/PitBoardKit/Support/ColorContrast.swift
import SwiftUI
import UIKit

/// Utilidades de contraste para pintar texto legible sobre colores de serie arbitrarios
/// (el usuario los elige libremente en el editor de series) — equivalente exacto de
/// `ColorContrast.kt`.
public enum ColorContrast {
    public static func perceivedLuminance(_ color: Color) -> Double {
        let (r, g, b, _) = components(of: color)
        return 0.299 * r + 0.587 * g + 0.114 * b
    }

    /// Empuja `color` hacia claro/oscuro si su luminancia percibida está demasiado cerca
    /// de la de `background`, para que no se confundan visualmente entre sí.
    public static func ensureContrast(_ color: Color, background: Color) -> Color {
        if abs(perceivedLuminance(color) - perceivedLuminance(background)) >= 0.25 { return color }

        var h: CGFloat = 0, s: CGFloat = 0, v: CGFloat = 0, a: CGFloat = 0
        _ = UIColor(color).getHue(&h, saturation: &s, brightness: &v, alpha: &a)

        let backgroundIsDark = perceivedLuminance(background) < 0.5
        let newValue = backgroundIsDark ? min(v + 0.4, 1.0) : max(v - 0.4, 0.0)
        return Color(UIColor(hue: h, saturation: s, brightness: newValue, alpha: a))
    }

    /// Blanco o negro, el que mejor se lea encima de `background`.
    public static func readableTextColor(background: Color) -> Color {
        perceivedLuminance(background) > 0.55 ? .black : .white
    }

    public static let fallbackColor = Color(hex: "#5F6570") ?? .gray

    public static func safeParseColor(_ hex: String, fallback: Color = ColorContrast.fallbackColor) -> Color {
        Color(hex: hex) ?? fallback
    }

    private static func components(of color: Color) -> (Double, Double, Double, Double) {
        var r: CGFloat = 0, g: CGFloat = 0, b: CGFloat = 0, a: CGFloat = 0
        _ = UIColor(color).getRed(&r, green: &g, blue: &b, alpha: &a)
        return (Double(r), Double(g), Double(b), Double(a))
    }
}

public extension Color {
    /// "#RRGGBB" o "#AARRGGBB" (con o sin "#") — equivalente de
    /// `android.graphics.Color.parseColor`. nil si el texto no es un hex válido, para que
    /// quien llama caiga a su propio color de respaldo (nunca un color inventado).
    init?(hex: String) {
        var s = hex.trimmingCharacters(in: .whitespacesAndNewlines)
        if s.hasPrefix("#") { s.removeFirst() }
        guard s.count == 6 || s.count == 8, let value = UInt64(s, radix: 16) else { return nil }

        let r, g, b, a: Double
        if s.count == 8 {
            a = Double((value >> 24) & 0xFF) / 255
            r = Double((value >> 16) & 0xFF) / 255
            g = Double((value >> 8) & 0xFF) / 255
            b = Double(value & 0xFF) / 255
        } else {
            a = 1
            r = Double((value >> 16) & 0xFF) / 255
            g = Double((value >> 8) & 0xFF) / 255
            b = Double(value & 0xFF) / 255
        }
        self.init(.sRGB, red: r, green: g, blue: b, opacity: a)
    }
}
